syntax = "proto3";

package DB.Manifest;

message UUIDInfo
{
  fixed64 low = 1;
  fixed64 high = 2;
}

enum DiskType {
  LOCAL = 0;
  REMOTE = 1;
}

message UINT128 {
    uint64 high64 = 1;
    uint64 low64 = 2;
}

message PartInfo
{
    string partition_id = 1;
    int64 min_block = 2;
    int64 max_block = 3;
    uint32 level = 4;
    int64 mutation = 5;
}

message SimpleChecksum {
    uint64 size = 2;
    UINT128 file_hash = 3;
    bool is_compressed = 4;
    uint64 uncompressed_size = 5;
    UINT128 uncompressed_hash = 6;
    bytes file_name = 7;
}

message SimpleChecksums {
    bool load_from_store = 1;
    int32 format_version = 2;
    repeated SimpleChecksum files = 5;
}

message Columns {
    bytes columns = 1;
    bytes serialization_infos = 2;
    int32 metadata_version = 3;
}

message IndexGranularityInfo {
   bool adaptive = 1;
   bool compressed = 2;
   uint64 fixed_index_granularity = 3;
   uint64 index_granularity_bytes = 4;
}

message IndexGranularity {
   IndexGranularityInfo info = 1;
   bool initialized = 2 [deprecated = true];
   repeated uint64 marks_rows_partial_sums = 3;
   optional bool is_constant = 4;
}

message TransactionID {
    uint64 start_csn = 1;
    uint64 local_tid = 2;
    UUIDInfo host_id = 3;
}

message VersionMetadata {
   TransactionID creation_tid = 1;
   TransactionID removal_tid = 2;
   uint64 removal_tid_lock = 3;
   uint64 creation_csn = 4;
   uint64 removal_csn = 5;
}

message PartObject {
    int32 version = 1;
    PartInfo info = 2;
    string path = 3;
    string name = 4;
    int32 type = 5;
    int32 state = 6;
    int64 bytes = 7;
    int64 rows = 8;
    UUIDInfo table_uuid = 10;
    UUIDInfo uuid = 11;
    uint64 last_modify_time = 12;
    bytes compress_codec = 13;
    SimpleChecksums checksums = 14;
    bytes partition = 15;
    Columns columns = 16;
    VersionMetadata version_metadata = 17;
    IndexGranularity index_granularity = 18;
    bytes minmax = 19;
    optional bytes ttl_infos = 20;
    optional string projection_name = 24;
    repeated PartObject projections = 25;
    optional bool primary_compressed = 26;
    optional bytes primary = 27;
    DiskType disk_type = 28;
    string disk_name = 29;
    string disk_path = 30;
}
